libtransact/src/database/sqlite.rs

```diff
@@ -28,42 +30,106 @@ use super::{
 
 type SqliteConnection = r2d2::PooledConnection<SqliteConnectionManager>;
 
-const DEFAULT_MMAP_SIZE: i64 = 100 * 1024 * 1024;
+/// The default size
+pub const DEFAULT_MMAP_SIZE: i64 = 100 * 1024 * 1024;
 
-#[derive(Clone)]
-pub struct SqliteDatabase {
-    pool: r2d2::Pool<SqliteConnectionManager>,
+/// A builder for generating SqliteDatabase instances.
+pub struct SqliteDatabaseBuilder {
+    path: Option<String>,
+    indexes: Vec<&'static str>,
+    pool_size: Option<u32>,
+    memory_map_size: i64,
 }
 
-impl SqliteDatabase {
-    pub fn new(path: &str, indexes: &[&str]) -> Result<Self, DatabaseError> {
-        eprintln!("new database at {}", path);
-        let manager = SqliteConnectionManager::file(path);
-        let pool = r2d2::Pool::new(manager).map_err(|err| {
-            DatabaseError::InitError(format!(
-                "unable to create sqlite connection pool to {}: {}",
-                path, err
-            ))
+impl SqliteDatabaseBuilder {
+    fn new() -> Self {
+        Self {
+            path: None,
+            indexes: vec![],
+            pool_size: None,
+            memory_map_size: DEFAULT_MMAP_SIZE,
+        }
+    }
+
+    /// Set the path for the Sqlite database.
+    pub fn with_path<S: Into<String>>(mut self, path: S) -> Self {
+        self.path = Some(path.into());
+        self
+    }
+
+    /// Set the names of the indexes to include in the database instance.
+    pub fn with_indexes(mut self, indexes: &[&'static str]) -> Self {
+        self.indexes = indexes.to_vec();
+        self
+    }
+
+    /// Set the connection pool size.
+    pub fn with_connection_pool_size(mut self, pool_size: NonZeroU32) -> Self {
+        self.pool_size = Some(pool_size.get());
+        self
+    }
+
+    /// Set the size used for Memory-Mapped I/O.
+    ///
+    /// This can be disabled by setting the value to `0`
```

Comment:
Could this instead by disabled using an option (i.e. `None` is disabled)?

Reply:
`0` is the sqlite configuration value, which I don't think we necessarily need to hide. If it was an option, then it would need to be `Option<NonZeroU64>`, which feels a little heavyweight for a builder. If it was `Option<u64>`, it would also be disableable via `Some(0)`.